UI/media-controls.cpp

```diff
@@ -0,0 +1,347 @@
+	ui->playPauseButton->style()->unpolish(ui->playPauseButton);
+	ui->playPauseButton->style()->polish(ui->playPauseButton);
+
+	ui->slider->setValue(0);
+	ui->timerLabel->setText("00:00:00");
+	ui->durationLabel->setText("00:00:00");
+	ui->slider->setEnabled(false);
+
+	StopTimer();
+}
+
+void MediaControls::RefreshControls()
+{
+	if (!source) {
+		SetRestartState();
+		setEnabled(false);
+		return;
+	} else {
+		setEnabled(true);
+	}
+
+	const char *id = obs_source_get_id(source);
+
+	if (id && *id && strcmp(id, "ffmpeg_source") == 0) {
+		ui->nextButton->hide();
+		ui->previousButton->hide();
+	} else {
+		ui->nextButton->show();
+		ui->previousButton->show();
+	}
+
+	obs_media_state state = obs_source_media_get_state(source);
+
+	switch (state) {
+	case OBS_MEDIA_STATE_STOPPED:
+	case OBS_MEDIA_STATE_ENDED:
+		SetRestartState();
+		break;
+	case OBS_MEDIA_STATE_PLAYING:
+		SetPlayingState();
+		break;
+	case OBS_MEDIA_STATE_PAUSED:
+		SetPausedState();
+		break;
+	default:
+		break;
+	}
+}
+
+OBSSource MediaControls::GetSource()
+{
+	return source;
+}
+
+void MediaControls::SetSource(OBSSource newSource)
+{
+	signal_handler_disconnect(obs_source_get_signal_handler(source),
+				  "media_play", OBSMediaPlay, this);
+	signal_handler_disconnect(obs_source_get_signal_handler(source),
+				  "media_pause", OBSMediaPause, this);
+	signal_handler_disconnect(obs_source_get_signal_handler(source),
+				  "media_restart", OBSMediaPlay, this);
+	signal_handler_disconnect(obs_source_get_signal_handler(source),
+				  "media_stopped", OBSMediaStopped, this);
+	signal_handler_disconnect(obs_source_get_signal_handler(source),
+				  "media_started", OBSMediaStarted, this);
+	signal_handler_connect(obs_source_get_signal_handler(source),
+			       "media_ended", OBSMediaStopped, this);
+
+	source = newSource;
+
+	if (source) {
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_play", OBSMediaPlay, this);
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_pause", OBSMediaPause, this);
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_restart", OBSMediaPlay, this);
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_stopped", OBSMediaStopped, this);
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_started", OBSMediaStarted, this);
+		signal_handler_connect(obs_source_get_signal_handler(source),
+				       "media_ended", OBSMediaStopped, this);
+	}
+
+	RefreshControls();
+}
+
+void MediaControls::SetSliderPosition()
+{
+	float time = (float)obs_source_media_get_time(source);
+	float duration = (float)obs_source_media_get_duration(source);
+
+	float sliderPosition = (time / duration) * (float)ui->slider->maximum();
+
+	ui->slider->setValue((int)sliderPosition);
+
+	ui->timerLabel->setText(FormatSeconds((int)(time / 1000.0f)));
+	ui->durationLabel->setText(FormatSeconds((int)(duration / 1000.0f)));
```

Comment:
Based on the requests we've been getting lately, I think people would rather see how much time is remaining in playback rather than the static duration of the media source, i.e. total duration minus time elapsed. A countdown time is important to see because people base manual cues off remaining time left in a media file, or do audible countdowns to other people on production. For example, they'll say things like "Camera 3 in 5...4...3...2...1...take" and base that countdown off the remaining time in playback.

Reply:
I actually would think that a elapsed/remaining format would be useful. Some things cue off either...